packages/common/src/hooks/chats/useAudiusLinkResolver.ts

```diff
@@ -0,0 +1,210 @@
+
+import {
+  AudiusSdk,
+  instanceOfPlaylist,
+  instanceOfTrack,
+  Playlist,
+  ResolveApi,
+  Track,
+  User
+} from '@audius/sdk'
+
+import { ID } from '~/models'
+import {
+  decodeHashId,
+  formatCollectionName,
+  formatTrackName,
+  formatUserName,
+  matchAudiusLinks,
+  Nullable
+} from '~/utils'
+
+const {
+  instanceOfTrackResponse,
+  instanceOfPlaylistResponse,
+  instanceOfUserResponse
+} = ResolveApi
+
+/**
+ * Hook to use within chat composition interfaces.
+ * It pulls Audius links out into display text
+ * @param value chat input text, e.g. what's in the text area
+ * @param hostname audius hostname for this instance, e.g. audius.co
+ * @param audiusSdk sdk instantiation interface
+ * @returns {Object} Object containing helper methods and state:
+ * - trackId: The ID of the first track resolved (or null if none found).
+ *
+ * - collectionId: The ID of the first collection resolved (or null if none found).
+ *
+ * - resolveLinks: Async function that takes the input text and
+ *     replaces Audius links with their human-readable counterparts.
+ *
+ * - restoreLinks: Function that restores human-readable text back into Audius
+ *      links in the provided value. This should be used before text submission.
+ *
+ * - getMatches: Function that returns a generator of matches for resolved
+ *      Audius links within the provided value. This is useful when displaying text
+ *      back to the user.
+ *
+ * - handleBackspace: Function that checks if the user backspaced over a resolved link
+ *      and modifies the text accordingly, deleting the whole display text/link on a
+ *      single keystroke.
+ *
+ * - reset: Function that resets the `trackId` and `collectionId` state to null.
+ *      Call this method when submitting.
+ */
+export const useAudiusLinkResolver = ({
+  value,
+  hostname,
+  audiusSdk
+}: {
+  value: string
+  hostname: string
+  audiusSdk: () => Promise<AudiusSdk>
+}) => {
+  // Maintain bidirectional maps of audius links to human readable format
+  const linkToHuman = useRef<{ [key: string]: string }>({}).current
+  const humanToData = useRef<{
+    [key: string]: { link: string; data: Track | Playlist | User }
+  }>({}).current
+
+  // The track id used to render the composer preview
+  const [trackId, setTrackId] = useState<Nullable<ID>>(null)
+  // The collection id used to render the composer preview
+  const [collectionId, setCollectionId] = useState<Nullable<ID>>(null)
+
+  /**
+   * Resolves Audius links in the provided text to their display text equivalent.
+   */
+  const resolveLinks = useCallback(
+    async (value: string) => {
+      const sdk = await audiusSdk()
+      const { matches } = matchAudiusLinks({
+        text: value,
+        hostname
+      })
+
+      for (const match of matches) {
+        if (!(match in linkToHuman)) {
+          const res = await sdk.resolve({ url: match })
+          if (res.data) {
+            if (instanceOfTrackResponse(res)) {
```

Comment:
love this much better. saw ur convo with marcus, need to change to typecheck data?

Reply:
need this here or we have to do the isArray checks here